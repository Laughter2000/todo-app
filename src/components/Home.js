import React, { useState, useEffect, useReducer, useRef } from "react";
import moon from "../images/icon-moon.svg";
import sun from "../images/icon-sun.svg";
import cross from "../images/icon-cross.svg";
import check from "../images/icon-check.svg";

export const todoReducer = (state, action) => {
  switch (action.type) {
    case "SAVED TODO":
      return [...state, ...action.payload];

    case "CREATE TODO":
      var newTodo = {
        item: action.payload,
        complete: false,
      };
      var duplicate = false;
      state.map((todo) => {
        if (todo.item === newTodo.item) {
          return (duplicate = true);
        } else {
          return duplicate;
        }
      });
      return duplicate ? state : [...state, newTodo];

    case "DO TODO":
      return state.map((todo) => {
        if (todo.item === action.payload) {
          return { ...todo, complete: true };
        } else {
          return todo;
        }
      });

    case "UNDO TODO":
      return state.map((todo) => {
        if (todo.item === action.payload) {
          return { ...todo, complete: false };
        } else {
          return todo;
        }
      });

    case "REMOVE TODO":
      return state.filter((todo) => todo.item !== action.payload);

    case "CLEAR COMPLETE":
      return state.filter((todo) => !todo.complete);

    default:
      return state;
  }
};
const Home = ({ change, mode }) => {
  const [todo, setTodo] = useState("");
  const [temp, setTemp] = useState([]);
  const [notification, setNotification] = useState("");

  const [todoList, dispatchTodo] = useReducer(todoReducer, []);
  const firstLoad = useRef(false);

  const submitTodo = (e) => {
    e.preventDefault();
    dispatchTodo({
      type: "CREATE TODO",
      payload: todo,
    });
  };

  const handleTodo = (todo) => {
    dispatchTodo({
      type: todo.complete ? "UNDO TODO" : "DO TODO",
      payload: todo.item,
    });
  };

  const filterTodo = (e) => {
    if (e === "All") {
      setTemp(todoList);
      setNotification("");
    } else if (e === "Completed") {
      setTemp(todoList.filter((todo) => todo.complete));
      setNotification("Completed");
    } else if (e === "Active") {
      setTemp(todoList.filter((todo) => !todo.complete));
      setNotification("Active");
    }
  };

  useEffect(() => {
    if (!firstLoad.current) {
      firstLoad.current = true;
      const total = localStorage.getItem("todoList");
      if (total) {
        dispatchTodo({ type: "SAVED TODO", payload: JSON.parse(total) });
        setTemp(todoList);
      }
    } else {
      localStorage.setItem("todoList", JSON.stringify(todoList));
      setTemp(todoList);
    }
  }, [todoList]);

  return (
    <div className={mode ? "container-fluid" : "container-fluid dark-mode"}>
      <div className={mode ? "cover-img" : "dark-img"}></div>
      <div className="container-div">
        <header className="header">
          <h1>TODO</h1>
          {mode ? (
            <img
              src={moon}
              alt="img mode"
              className="img-mode"
              onClick={change}
            />
          ) : (
            <img
              src={sun}
              alt="img mode"
              className="img-mode"
              onClick={change}
            />
          )}
        </header>

        <main>
          <form onSubmit={submitTodo}>
            <div className={mode ? "input-div" : "input-div background-dark"}>
              <div className="circle" onClick={submitTodo}></div>
              <input
                type="text"
                className={
                  mode ? "input-tab" : "input-tab background-dark dark-text"
                }
                placeholder="Create a new todo..."
                onChange={(e) => setTodo(e.target.value)}
                onBlur={(e) => setTodo(e.target.value)}
              />
            </div>
          </form>

          <div className={mode ? "todo-list" : "todo-list background-dark"}>
            {(() => {
              if (temp.length >= 1) {
                return (
                  <div>
                    {temp.map((todo) => (
                      <div className="todo-item" key={todo.item}>
                        <div className="content">
                          <div
                            className={
                              todo.complete ? "circle-colored" : "circle"
                            }
                            onClick={() => handleTodo(todo)}
                          >
                            {todo.complete ? (
                              <img
                                src={check}
                                alt="cross"
                                className="icon-check"
                              />
                            ) : (
                              " "
                            )}
                          </div>
                          <p
                            className={
                              todo.complete ? "todo-p completed-p" : "todo-p"
                            }
                          >
                            {todo.item}
                          </p>
                        </div>
                        <img
                          src={cross}
                          alt="cross"
                          className="cross"
                          onClick={() =>
                            dispatchTodo({
                              type: "REMOVE TODO",
                              payload: todo.item,
                            })
                          }
                        />
                      </div>
                    ))}
                    <div className="menu">
                      <p>
                        {todoList.length >= 1
                          ? todoList.length -
                            todoList.filter((todo) => todo.complete).length
                          : 0}{" "}
                        items left
                      </p>
                      <div>
                        <p
                          className={
                            notification === "All" || notification === ""
                              ? "menu-toggle filter"
                              : "menu-toggle"
                          }
                          onClick={() => filterTodo("All")}
                        >
                          All
                        </p>
                        <p
                          className={
                            notification === "Active"
                              ? "menu-toggle filter"
                              : "menu-toggle"
                          }
                          onClick={() => filterTodo("Active")}
                        >
                          Active
                        </p>
                        <p
                          className={
                            notification === "Completed"
                              ? "menu-toggle filter"
                              : "menu-toggle"
                          }
                          onClick={() => filterTodo("Completed")}
                        >
                          Completed
                        </p>
                      </div>
                      <p
                        className="last"
                        onClick={() =>
                          dispatchTodo({
                            type: "CLEAR COMPLETE",
                          })
                        }
                      >
                        Clear Completed
                      </p>
                    </div>
                  </div>
                );
              } else {
                return (
                  <div>
                    <p className="empty">No {notification} todo item yet</p>
                    <div className="menu">
                      <p>
                        {todoList.length >= 1
                          ? todoList.length -
                            todoList.filter((todo) => todo.complete).length
                          : 0}{" "}
                        items left
                      </p>
                      <div>
                        <p
                          className={
                            notification === "All" || notification === ""
                              ? "menu-toggle filter"
                              : "menu-toggle"
                          }
                          onClick={() => filterTodo("All")}
                        >
                          All
                        </p>
                        <p
                          className={
                            notification === "Active"
                              ? "menu-toggle filter"
                              : "menu-toggle"
                          }
                          onClick={() => filterTodo("Active")}
                        >
                          Active
                        </p>
                        <p
                          className={
                            notification === "Completed"
                              ? "menu-toggle filter"
                              : "menu-toggle"
                          }
                          onClick={() => filterTodo("Completed")}
                        >
                          Completed
                        </p>
                      </div>
                      <p
                        className="last"
                        onClick={() =>
                          dispatchTodo({
                            type: "CLEAR COMPLETE",
                          })
                        }
                      >
                        Clear Completed
                      </p>
                    </div>
                  </div>
                );
              }
            })()}
          </div>
          <div className={mode ? "second-menu" : "second-menu dark-one"}>
            <p
              onClick={() => filterTodo("All")}
              className={
                notification === "All" || notification === "" ? "filter" : ""
              }
            >
              All
            </p>
            <p
              onClick={() => filterTodo("Active")}
              className={notification === "Active" ? "filter" : ""}
            >
              Active
            </p>
            <p
              onClick={() => filterTodo("Completed")}
              className={notification === "Completed" ? "filter" : ""}
            >
              Completed
            </p>
          </div>
        </main>
      </div>
    </div>
  );
};

export default Home;
