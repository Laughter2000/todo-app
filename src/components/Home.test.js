import render from "react-test-renderer";
import { mount } from "enzyme";
import Home, { todoReducer } from "./Home";

const todoList = [
  {
    item: "Learn more of React",
    complete: false,
  },
  {
    item: "Practice more with Frontend mentor challenges",
    complete: true,
  },
  {
    item: "Learn Redux",
    complete: false,
  },
];
describe("Home", () => {
  describe("Reducer", () => {
    it("should set todo list", () => {
      const state = [];
      const newState = todoReducer(state, {
        type: "SAVED TODO",
        payload: todoList,
      });
      expect(newState.length).toEqual(3);
    });

    it("should not create a duplicate todo item", () => {
      const state = todoList;
      const newState = todoReducer(state, {
        type: "CREATE TODO",
        payload: "Learn Redux",
      });
      expect(newState.length).toEqual(3);
    });

    it("should a unique todo item", () => {
      const state = todoList;
      const newState = todoReducer(state, {
        type: "CREATE TODO",
        payload: "Learn Firebase",
      });
      expect(newState.length).toEqual(4);
    });

    it("should set todo item to completed", () => {
      const state = todoList;
      const newState = todoReducer(state, {
        type: "UNDO TODO",
        payload: todoList[1].item,
      });
      expect(newState[1].complete).toBeFalsy;
    });
  });

  it("should set todo item as not completed", () => {
    const state = todoList;
    const newState = todoReducer(state, {
      type: "DO TODO",
      payload: todoList[0].item,
    });
    expect(newState[0].complete).toBeTruthy;
  });

  it("should remove a todo item", () => {
    const state = todoList;
    const newState = todoReducer(state, {
      type: "REMOVE TODO",
      payload: todoList[2].item,
    });
    expect(newState.length).toEqual(2);
  });

  it("should clear completed", () => {
    const state = todoList;
    const newState = todoReducer(state, {
      type: "CLEAR COMPLETE",
    });
    expect(newState.length).toEqual(2);
  });

  test("Has a valid snap short", () => {
    const component = render.create(<Home />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
