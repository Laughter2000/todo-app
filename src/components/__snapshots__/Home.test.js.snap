// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Has a valid snap short 1`] = `
<div
  className="container-fluid dark-mode"
>
  <div
    className="dark-img"
  />
  <div
    className="container-div"
  >
    <header
      className="header"
    >
      <h1>
        TODO
      </h1>
      <img
        alt="img mode"
        className="img-mode"
        src="icon-sun.svg"
      />
    </header>
    <main>
      <form
        onSubmit={[Function]}
      >
        <div
          className="input-div background-dark"
        >
          <div
            className="circle"
            onClick={[Function]}
          />
          <input
            className="input-tab background-dark dark-text"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Create a new todo..."
            type="text"
          />
        </div>
      </form>
      <div
        className="todo-list background-dark"
      >
        <div>
          <p
            className="empty"
          >
            No 
            
             todo item yet
          </p>
          <div
            className="menu"
          >
            <p>
              0
               
              items left
            </p>
            <div>
              <p
                className="menu-toggle filter"
                onClick={[Function]}
              >
                All
              </p>
              <p
                className="menu-toggle"
                onClick={[Function]}
              >
                Active
              </p>
              <p
                className="menu-toggle"
                onClick={[Function]}
              >
                Completed
              </p>
            </div>
            <p
              className="last"
              onClick={[Function]}
            >
              Clear Completed
            </p>
          </div>
        </div>
      </div>
      <div
        className="second-menu dark-one"
      >
        <p
          className="filter"
          onClick={[Function]}
        >
          All
        </p>
        <p
          className=""
          onClick={[Function]}
        >
          Active
        </p>
        <p
          className=""
          onClick={[Function]}
        >
          Completed
        </p>
      </div>
    </main>
  </div>
</div>
`;
